src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,173 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node newNode = new Node(null, value, null);
+        if (isEmpty()) {
+            last = newNode;
+            first = newNode;
+        } else {
+            newNode.prev = last;
+            last.next = newNode;
+            last = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, size);
+        Node newNode = new Node(null, value, null);
+        if (first == null || index == size) {
+            add(value);
+        } else if (index == 0) {
+            newNode.next = first;
+            first.prev = newNode;
+            first = newNode;
+            size++;
+        } else {
+            Node<T> current = getNodeByIndex(index - 1);
+            newNode.next = current.next;
+            current.next.prev = newNode;
+            current.next = newNode;
+            newNode.prev = current;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size - 1);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, size - 1);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size - 1);
```

Comment:
```suggestion
        checkIndex(index);
```

Reply:
you have access to size from any place in this class

Reply:
But different methods need different size limit, so I pass checkIndex(int index, int size). 
in void add(T value, int index) method I check checkIndex(index, size), 
and in get(int index) I check checkIndex(index, size - 1);

Reply:
better to create two check methods for such cases